src/mpid/ch4/netmod/ofi/ofi_init.h

```diff
@@ -624,23 +629,39 @@ static inline int MPIDI_NM_mpi_init_hook(int rank,
         sprintf(keyS, "OFI-%d", rank);
         MPIDI_OFI_PMI_CALL_POP(PMI_KVS_Put(MPIDI_Global.kvsname, keyS, val), pmi);
         MPIDI_OFI_PMI_CALL_POP(PMI_KVS_Commit(MPIDI_Global.kvsname), pmi);
-        MPIDI_OFI_PMI_CALL_POP(PMI_Barrier(), pmi);
+
+        for (i = 0; i < size; i++) {
+            if (MPIDI_CH4_rank_is_local(i, MPIR_Process.comm_world)) {
+                if (i == rank)
+                    local_rank = num_local;
+
+                if (local_rank_0 == -1)
+                    local_rank_0 = i;
+
+                num_local++;
+            }
+        }
+
+        MPIDU_shm_seg_alloc(size * MPIDI_Global.addrnamelen, (void **)&table);
+        MPIDU_shm_seg_commit(&memory, &barrier, num_local, local_rank, local_rank_0, rank);
 
         /* -------------------------------- */
         /* Create our address table from    */
         /* encoded KVS values               */
         /* -------------------------------- */
-        table = (char *) MPL_malloc(size * MPIDI_Global.addrnamelen);
-        maxlen = MPIDI_KVSAPPSTRLEN;
-
-        for (i = 0; i < size; i++) {
+        start = local_rank * (size / num_local);
+        end = start + (size / num_local);
+        if (local_rank == num_local - 1)
+            end += size % num_local;
+        for (i = start; i < end; i++) {
             sprintf(keyS, "OFI-%d", i);
             MPIDI_OFI_PMI_CALL_POP(PMI_KVS_Get
                                    (MPIDI_Global.kvsname, keyS, valS, MPIDI_KVSAPPSTRLEN), pmi);
             MPIDI_OFI_STR_CALL(MPL_str_get_binary_arg
                                (valS, "OFI", (char *) &table[i * MPIDI_Global.addrnamelen],
                                 MPIDI_Global.addrnamelen, &maxlen), buscard_len);
         }
+        PMI_Barrier();
```

Comment:
Ok, that sounds good then. There probably isn't much drift at this point so as long as the barrier algorithm is cheap then things shouldn't be too expensive.

On a related note, it would be cool if PMI(x) could hook into OFI to enable accelerated communication. I'm sure that would be a pretty major redesign, but it would potentially make a big difference for these kinds of operations.